### PR TITLE
Include policies for securing sandbox namespace 

### PR DESCRIPTION
**Pull request by CosimoMichelagnoli**
# Description

This PR aims to apply some cluster-wide policies based on Kyverno framework. Three main security requirements: 
- Avoid creation of privileged pods
- Avoid creation of load balancer services (i.e avoid creation of routable IPs) and nodePort
- Force specific name for ingress hostname

These policies are applied only to sandbox namespaces.

### CODE REVIEW

**Review by giorio94 — COMMENTED** [review]
A few initial comments inline. Additionally:

* For the moment, I would personally drop the `crownlabs-policies-privilege-escalation` and `crownlabs-policies-run-as-non-root` policies, as they make sense, but require a modification of the deployment manifests (this could create problems when students experiment for the first time).
* I would add the following policies (from the kyverno repo), as targeting definitely dangerous aspects:
  - [Disallow Host Namespaces](https://kyverno.io/policies/pod-security/baseline/disallow-host-namespaces/disallow-host-namespaces/)
  - [Disallow hostPath](https://kyverno.io/policies/pod-security/baseline/disallow-host-path/disallow-host-path/)
  - [Disallow hostPorts](https://kyverno.io/policies/pod-security/baseline/disallow-host-ports/disallow-host-ports/)
* I would keep the information about the installation of kyverno in the `infrastructure/policies` folder, while moving the chart with the policies in a `policies` folder in the root of the repo, to separate the two things (as we will also need to customize the kyverno values file when installing it on CrownLabs).

**Comment by CosimoMichelagnoli** [comment]
> A few initial comments inline. Additionally:
> 
> * For the moment, I would personally drop the `crownlabs-policies-privilege-escalation` and `crownlabs-policies-run-as-non-root` policies, as they make sense, but require a modification of the deployment manifests (this could create problems when students experiment for the first time).
> * I would add the following policies (from the kyverno repo), as targeting definitely dangerous aspects:
>   
>   * [Disallow Host Namespaces](https://kyverno.io/policies/pod-security/baseline/disallow-host-namespaces/disallow-host-namespaces/)
>   * [Disallow hostPath](https://kyverno.io/policies/pod-security/baseline/disallow-host-path/disallow-host-path/)
>   * [Disallow hostPorts](https://kyverno.io/policies/pod-security/baseline/disallow-host-ports/disallow-host-ports/)
> * I would keep the information about the installation of kyverno in the `infrastructure/policies` folder, while moving the chart with the policies in a `policies` folder in the root of the repo, to separate the two things (as we will also need to customize the kyverno values file when installing it on CrownLabs).

We just corrected the PR following your comments. Hoping to have done everything right, how should we proceed with regards to the kyverno values ​​to be customized?

**Review by giorio94 — COMMENTED** [review]
A couple of additional comments inline. Moreover, please delete the empty values file for the moment (the one in the infrastructure folder).

File: policies/templates/disallow-hostPath.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    kyverno.io/kubernetes-version: "1.22-1.23"
+    policies.kyverno.io/description: >-
+      HostPath volumes let Pods use host directories and volumes in containers.
+      Using host resources can be used to access shared data or escalate privileges
+      and should not be allowed. This policy ensures no hostPath volumes are in use.      
+spec:
+  validationFailureAction: enforce
+  background: false
+  rules:
+    - name: host-path
```
Comment (giorio94) [inline policies/templates/disallow-hostPath.yaml:22]:
```suggestion
    - name: disallow-host-path-volumes
```

File: policies/templates/disallow-hostPorts.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    kyverno.io/kubernetes-version: "1.22-1.23"
+    policies.kyverno.io/description: >-
+      Access to host ports allows potential snooping of network traffic and should not be
+      allowed, or at minimum restricted to a known list. This policy ensures the `hostPort`
+      field is unset or set to `0`.       
+spec:
+  validationFailureAction: enforce
+  background: false
+  rules:
+    - name: host-ports-none
```
Comment (giorio94) [inline policies/templates/disallow-hostPorts.yaml:22]:
```suggestion
    - name: disallow-host-ports
```

File: policies/templates/disallow-loadbalancer.yaml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      Especially in cloud provider environments, a Service having type LoadBalancer will cause the
+      provider to respond by creating a load balancer somewhere in the customer account. This adds
+      cost and complexity to a deployment. Without restricting this ability, users may easily
+      overrun established budgets and security practices set by the organization. This policy restricts
+      use of the Service type LoadBalancer.      
+spec:
+  validationFailureAction: enforce
+  background: false
+  rules:
+  - name: no-LoadBalancer
```
Comment (giorio94) [inline policies/templates/disallow-loadbalancer.yaml:22]:
```suggestion
  - name: disallow-loadbalancer-services
```

File: policies/templates/disallow-nodeport.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      A Kubernetes Service of type NodePort uses a host port to receive traffic from
+      any source. A NetworkPolicy cannot be used to control traffic to host ports.
+      Although NodePort Services can be useful, their use must be limited to Services
+      with additional upstream security checks. This policy validates that any new Services
+      do not use the `NodePort` type.          
+spec:
+  validationFailureAction: enforce
+  background: false
+  rules:
+  - name: no-Nodeport
```
Comment (giorio94) [inline policies/templates/disallow-nodeport.yaml:22]:
```suggestion
  - name: disallow-nodeport-services
```

File: policies/templates/disallow-privileged-containers.yaml [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    kyverno.io/kyverno-version: 1.6.0
+    kyverno.io/kubernetes-version: "1.22-1.23"
+    policies.kyverno.io/description: >-
+      Privileged mode disables most security mechanisms and must not be allowed. This policy
+      ensures Pods do not call for privileged mode.          
+spec:
+  validationFailureAction: enforce
+  background: false
+  rules:
+    - name: privileged-containers
```
Comment (giorio94) [inline policies/templates/disallow-privileged-containers.yaml:21]:
```suggestion
    - name: disallow-privileged-containers
```

File: policies/templates/restrict-host-ingress.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    policies.kyverno.io/category: Security
+    policies.kyverno.io/subject: Ingress
+    policies.kyverno.io/description: >-
+        An ingress resource needs to define a valid hostname. 
+        This policy ensures that the hostname follow the form sxxxxxx.sandbox.crownlabs.polito.it.
+spec:
+  validationFailureAction: enforce
+  background: false
+  rules:
+  - name: check-for-host
```
Comment (giorio94) [inline policies/templates/restrict-host-ingress.yaml:18]:
```suggestion
  - name: restrict-ingress-hostnames
```

File: policies/templates/disallow-hostPath.yaml [hunk 7]
```diff
@@ -0,0 +1,38 @@
+apiVersion: kyverno.io/v1
```
Comment (giorio94) [inline policies/templates/disallow-hostPath.yaml:1]: Please, keep filenames lowercase (same for other files).

File: policies/templates/restrict-host-ingress.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: {{ include "policies.fullname" . }}-check-for-host
```
Comment (giorio94) [inline policies/templates/restrict-host-ingress.yaml:4]:
```suggestion
  name: {{ include "policies.fullname" . }}-restrict-host-ingress
```

Please, ensure that the policy name matches that of the file. Same for all others.

**Review by giorio94 — APPROVED** [review]
A couple of final nits inline. Please, generalize also the commit message, then we can go ahead and merge the PR.

**Pull request merged**